File: frontend/src/core/domain/Friendship.ts
// フラットなFriendship（フレンドリスト用）
export interface FriendInfo {
  friendship: {
    id: string;
    requester_id: string;
    addressee_id: string;
    status: 'pending' | 'accepted' | 'rejected' | 'blocked';
    created_at: string;
    updated_at: string;
  };
  friend: {
    id: string;
    username: string;
    display_name: string;
    avatar_url?: string;
  };
}

// 保留中申請（リクエスト用）
export interface PendingRequestInfo {
  friendship: {
    id: string;
    requester_id: string;
    addressee_id: string;
    status: 'pending' | 'accepted' | 'rejected' | 'blocked';
    created_at: string;
    updated_at: string;
  };
  requester: {
    id: string;
    username: string;
    display_name: string;
    avatar_url?: string;
  };
}

// 後方互換のためFriendship型も維持
export interface Friendship {
  id: string;
  requester_id: string;
  addressee_id: string;
  status: 'pending' | 'accepted' | 'rejected' | 'blocked';
  created_at: string;
  updated_at: string;
  requester?: {
    id: string;
    username: string;
    display_name: string;
    avatar_url?: string;
  };
  addressee?: {
    id: string;
    username: string;
    display_name: string;
    avatar_url?: string;
  };
}

export interface FriendRequestRequest {
  addressee_id: string;
}

export interface FriendActionRequest {
  friendship_id: string;
}

File: backend/entities/friendship.go
package entities

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

// FriendshipStatus は友達関係のステータス
type FriendshipStatus string

const (
	FriendshipStatusPending  FriendshipStatus = "pending"
	FriendshipStatusAccepted FriendshipStatus = "accepted"
	FriendshipStatusRejected FriendshipStatus = "rejected"
	FriendshipStatusBlocked  FriendshipStatus = "blocked"
)

// Friendship は友達関係エンティティ
type Friendship struct {
	ID          uuid.UUID
	RequesterID uuid.UUID        // 申請者
	AddresseeID uuid.UUID        // 受信者
	Status      FriendshipStatus
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// NewFriendship は新しい友達申請を作成
func NewFriendship(requesterID, addresseeID uuid.UUID) (*Friendship, error) {
	if requesterID == addresseeID {
		return nil, errors.New("cannot send friend request to yourself")
	}

	return &Friendship{
		ID:          uuid.New(),
		RequesterID: requesterID,
		AddresseeID: addresseeID,
		Status:      FriendshipStatusPending,
		CreatedAt:   time.Now(),
		UpdatedAt:   time.Now(),
	}, nil
}

// Accept は友達申請を承認
func (f *Friendship) Accept() error {
	if f.Status != FriendshipStatusPending {
		return errors.New("friendship is not in pending status")
	}
	f.Status = FriendshipStatusAccepted
	f.UpdatedAt = time.Now()
	return nil
}

// Reject は友達申請を拒否
func (f *Friendship) Reject() error {
	if f.Status != FriendshipStatusPending {
		return errors.New("friendship is not in pending status")
	}
	f.Status = FriendshipStatusRejected
	f.UpdatedAt = time.Now()
	return nil
}

// Block はユーザーをブロック
func (f *Friendship) Block() {
	f.Status = FriendshipStatusBlocked
	f.UpdatedAt = time.Now()
}

// IsAccepted は友達関係が承認済みかどうかを確認
func (f *Friendship) IsAccepted() bool {
	return f.Status == FriendshipStatusAccepted
}
